### PR TITLE
Deprecation:Tracer#configure

### PR DESCRIPTION
**Pull request by marcotc**
This PR deprecates the `Datadog::Tracer#configure` method, which partially reconfigured the live tracer at any point of the host application execution. This flexibility created many complications, as the trace writer and sampler could be changed at any point in time, creating possibility for race conditions in many points of interactions of the internal `ddtrace` components.

The functionality from `Datadog::Tracer#configure` has been migrated to `Datadog::Tracer#new`.
Because `#configure` was invoked in the regular `ddtrace` initialization (at [`Components#build_tracer`](https://github.com/DataDog/dd-trace-rb/blob/148dfaeac9a890d8d9ec4fd74abfd3621869f33c/lib/ddtrace/configuration/components.rb#L71-L76)), there was quite a bit of refactoring to do to consolidate all tracer initialization into `Tracer#new`.

The main changes are the decoupling of the Writer and Sampler initialization from the Tracer class.
In today's state, Writer and Sampler would ideally be their own components as they are fully configurable not strictly dependent on the Tracer. The initialization logic for the Writer and Sampler has been extracted (now living in `Components`), but I deemed promoting them to their own components as out of scope for this PR, given how many changes are already in place here. I left comments in our source code around the future of these components and a rationale for it.

One noteworthy issue that was tackled is the coupling between Writer and Sampler today: the Writer updates the Sampler's service rates, if the sampler is a `PrioritySampler`. This wiring was done by the Tracer, but this has now been extracted: the Writer emits events, to which a subscriber listens to and updates the sampler if needed.

All `ddtrace` behaviour, except for the deprecation of `Tracer#configure`, should remain the same after this change.

### CODE REVIEW

File: lib/ddtrace/configuration/components.rb [hunk 1]
```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
...
+            Datadog::ContextFlush::Finished.new
+          end
+        end
+
+        # TODO: Sampler should be a top-level component.
+        # It is currently part of the Tracer initialization
+        # process, but can take a variety of options (including
+        # a fully custom instance) that makes the Tracer
+        # initialization process complex.
+        def build_sampler(settings)
```
Comment (ivoanjo) [inline lib/ddtrace/configuration/components.rb:93]: I'm curious about this comment -- what would be the use of an independent sampler that exists separately from a tracer? (Same question for the writer)

File: lib/ddtrace/configuration/components.rb [hunk 2]
```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
...
+            return writer
+          end
+
+          Writer.new(agent_settings: agent_settings, **settings.tracer.writer_options)
+        end
+
+        def subscribe_to_writer_events!(writer, sampler)
+          return unless writer.respond_to?(:events) # Check if it's a custom, external writer
+
+          writer.events.after_send.subscribe(:record_environment_information, &WRITER_RECORD_ENVIRONMENT_INFORMATION_CALLBACK)
```
Comment (ivoanjo) [inline lib/ddtrace/configuration/components.rb:143]: Minor: I would suggest maybe directly inlining this, as I think it ends up being harder to read/follow than a regular block in here.

**Review by ivoanjo — COMMENTED** [review]
Apologies for the delay on this PR, I've been kinda leaving it for last as it needs a bit of thinking, and I haven't yet been able to get through it completely.

I've left a few comments, and I'll come back to it tomorrow :)

File: lib/ddtrace/configuration/components.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/configuration/components.rb:93]: Yeah, something like that.
The sampler object has so many complex options (it's basically a rule engine), that it made the tracer configuration too heavy.

File: lib/ddtrace/configuration/components.rb (hunk 2 above)
Comment (marcotc) [reply lib/ddtrace/configuration/components.rb:143]: I had this inlined at first, but then I couldn't test what's inside `WRITER_RECORD_ENVIRONMENT_INFORMATION_CALLBACK` and `writer_update_priority_sampler_rates_callback` without making the `subscribe_to_writer_events!` test an integration test. The integration test looked super ugly, with too many options/combinations.

File: lib/ddtrace/configuration/components.rb (hunk 1 above)
Comment (ivoanjo) [reply lib/ddtrace/configuration/components.rb:93]: Hmm, I guess I see what you mean in terms of complexity, but I guess part of my confusion comes from what _being top-level_ means; in particular, and looking at the code below:

```ruby
      attr_reader \
        :health_metrics,
        :logger,
        :profiler,
        :runtime_metrics,
        :tracer

      def initialize(settings)
        # Logger
        @logger = self.class.build_logger(settings)

        agent_settings = AgentSettingsResolver.call(settings, logger: @logger)

        # Tracer
        @tracer = self.class.build_tracer(settings, agent_settings)

        # Profiler
        @profiler = self.class.build_profiler(settings, agent_settings, @tracer)

        # Runtime metrics
        @runtime_metrics = self.class.build_runtime_metrics_worker(settings)

        # Health metrics
        @health_metrics = self.class.build_health_metrics(settings)
      end
```

...when you mean "top-level", do you mean having `@sampler = ...` and `@writer = ...` here as well?

Because it doesn't seem quite correct to me, since while they are quite big/complex, they still seem like implementation details of tracing...?

File: lib/ddtrace/configuration/components.rb [hunk 3]
```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
...
+            context_flush = build_test_mode_context_flush(settings)
+            sampler = build_test_mode_sampler
+            writer = build_test_mode_writer(settings, agent_settings)
+          else
+            context_flush = build_context_flush(settings)
+            sampler = build_sampler(settings)
+            writer = build_writer(settings, agent_settings)
+          end
+
+          subscribe_to_writer_events!(writer, sampler)
```
Comment (ivoanjo) [inline lib/ddtrace/configuration/components.rb:68]: To be honest, I had left here a comment suggesting that some of this behavior be moved to the `Sampler` or the `Writer`, and then reading below I realized that this PR actually is going in the other direction :)

I'm... not entirely convinced lol. The interactions between tracer, sampler and writer are already complex enough, that I think this abstraction makes the whole thing even more hard to follow.

Would you consider breaking this change off from this PR and into its own? That would allow us to focus on that interaction separately from rewiring how all three classes get initialized.

File: lib/ddtrace/configuration/components.rb [hunk 4]
```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
...
-          #       existing mutable #configure function. Remove when these components
-          #       are extracted.
-          tracer.configure(agent_settings: agent_settings, **build_tracer_options(settings, agent_settings))
+        # Create new lambda for writer callback,
+        # capture the current sampler in the callback closure.
+        def writer_update_priority_sampler_rates_callback(sampler)
+          lambda do |_, responses|
+            response = responses.last
 
-          tracer
```
Comment (ivoanjo) [inline lib/ddtrace/configuration/components.rb:78]: I really like the removal of the two-step (init + configure) initialization of the Tracer!

File: lib/ddtrace/configuration/components.rb [hunk 5]
```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
...
+        end
+
+        def ensure_priority_sampling(sampler = nil)
+          if sampler.is_a?(PrioritySampler)
+            sampler
+          else
+            PrioritySampler.new(
+              base_sampler: sampler,
+              post_sampler: Sampling::RuleSampler.new
+            )
```
Comment (ivoanjo) [inline lib/ddtrace/configuration/components.rb:123]: I think the `RuleSampler.new` call on line 122 also should get the `rate_limit` and `default_sample_rate` arguments from `settings`, as otherwise it's a bit weird that on some codepaths we use the `settings` argument, and on others, the instance directly reaches in `Datadog.configuration` (like that class is doing).

Once we do so, we should have three initializations of `RuleSampler` that look exactly the same, so we get a free cleanup :)

Finally -- does it really make sense to have the `RuleSampler` class by default reach into `Datadog.configuration` like it is doing right now? Would it make sense for 1.0 to just make `rate_limit:` and `default_sample_rate:` required arguments with no defaults instead? (I guess I don't quite understand if that class is expected to be usable by customers directly, and if in that case it makes sense to have the nice defaults).

File: lib/ddtrace/context_flush.rb [hunk 6]
```diff
@@ -26,6 +26,8 @@ class Partial
       # Start flushing partial trace after this many active spans in one trace
       DEFAULT_MIN_SPANS_FOR_PARTIAL_FLUSH = 500
 
+      attr_reader :min_spans_for_partial
+
```
Comment (ivoanjo) [inline lib/ddtrace/context_flush.rb:30]: Hmmm, is this API being made public only for the tests, or are customers expected to use it as well?

I guess this is one of the challenges we face today -- I think we kinda expose a tad too much that it's not clear if it's just because we need to validate it or because it's useful for customers 😅

File: lib/ddtrace/tracer.rb [hunk 7]
```diff
@@ -63,61 +63,16 @@ def call_context(key = nil)
     # * +enabled+: set if the tracer submits or not spans to the local agent. It's enabled
     #   by default.
     def initialize(options = {})
```
Comment (ivoanjo) [inline lib/ddtrace/tracer.rb:65]: Minor: Since we're massively changing this method, perhaps a good time to move to real kw args? :)

File: lib/ddtrace/tracer.rb [hunk 8]
```diff
@@ -63,61 +63,16 @@ def call_context(key = nil)
...
-                         else
-                           Datadog::ContextFlush::Finished.new
-                         end
-      end
+      @sampler = options[:sampler] || PrioritySampler.new(
+        base_sampler: Datadog::AllSampler.new,
+        post_sampler: Sampling::RuleSampler.new,
+      )
+      @tags = options[:tags] || {}
+      @writer = options[:writer] || Datadog::Writer.new
```
Comment (ivoanjo) [inline lib/ddtrace/tracer.rb:75]: Since this class is rather complex, I don't think we should try to somewhat duplicate its initialization logic here. E.g. `Tracer.new` should come from `#build_tracer` in the components; if we would like to have that be standalone, maybe `#build_tracer` should be moved to another class (?).

I don't see a very interesting use case for supporting `Tracer.new` and so having all these defaults seems a bit unneeded.

**Review by ivoanjo — COMMENTED** [review]
Ok, finally went through it all!

I think we're moving in the right direction, and most of my feedback is local. Let me know your thoughts! :)

File: lib/ddtrace/configuration/components.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/configuration/components.rb:93]: Because they can be completely overwritten by the user with custom instances, it makes sense for them to be easy to track, and not overly attached to our tracing internals.
They are still tracer specific.
Maybe they could live under `Components::Tracing#writer` instead of `Components#writer`.

File: lib/ddtrace/configuration/components.rb (hunk 5 above)
Comment (marcotc) [reply lib/ddtrace/configuration/components.rb:123]: > I think the RuleSampler.new call on line 122 also should get the rate_limit and default_sample_rate arguments from settings, as otherwise it's a bit weird that on some codepaths we use the settings argument, and on others, the instance directly reaches in Datadog.configuration (like that class is doing).

Good call.

> Finally -- does it really make sense to have the RuleSampler class by default reach into Datadog.configuration like it is doing right now? Would it make sense for 1.0 to just make rate_limit: and default_sample_rate: required arguments with no defaults instead? (I guess I don't quite understand if that class is expected to be usable by customers directly, and if in that case it makes sense to have the nice defaults).

The issue is that we told plenty of clients to use:
```ruby
Datadog.configure do |c|
  c.tracer sampler: Datadog::PrioritySampler.new(
    post_sampler: Datadog::Sampling::RuleSampler.new
  )
end
```
to configure [TwL](https://docs.datadoghq.com/tracing/trace_search_and_analytics/#tracing-without-limits-recommended). And this API, without explicit constructions parameters was a product requirement at the time. It's  not great in hindsight, but I fear it would break too many configuration setups today.

File: lib/ddtrace/context_flush.rb (hunk 6 above)
Comment (marcotc) [reply lib/ddtrace/context_flush.rb:30]: It's for testing.
I can make it private.
I don't think we have the right tools in Ruby sometimes for testing.

File: lib/ddtrace/tracer.rb (hunk 7 above)
Comment (marcotc) [reply lib/ddtrace/tracer.rb:65]: I didn't want to, to not disturb Ruby 3.0 users, but I'm easily convinced. Made this change!

File: lib/ddtrace/configuration/components.rb (hunk 3 above)
Comment (marcotc) [reply lib/ddtrace/configuration/components.rb:68]: > Would you consider breaking this change off from this PR and into its own? That would allow us to focus on that interaction separately from rewiring how all three classes get initialized.

Break off the initialization changes of the Writer and Sampler from removing `Tracer#configure`?

I could do the initialization changes of the Writer and Sampler before removing `Tracer#configure`, but not in parallel. They would also overlap a bit: I'd have to write a stop-gap solution for the `Writer#update_priority_sampler` change, reconciling the old way of providing the `priority_sampler` to the Writer, while maintaining it at `Tracer#configure` and write tests to account for both cases and their conflicts, and then remove that stop-gap solution in the `Tracer#configure` PR.

If I break it down like suggested, it would make reviewing easier (unless the stop gap solutions turn out uglier and make reviewing the intermediate PRs worse), but it won't affect the end result. Given I can't even think of an easy way to gracefully break down the `Writer.@priority_sampler` dependency without making pretty much most of the changes in this PR, I'm inclined to say that I don't think breaking it down like suggested is a good idea.

File: lib/ddtrace/tracer.rb (hunk 8 above)
Comment (marcotc) [reply lib/ddtrace/tracer.rb]: @delner I'm torn on removing the ability to build a brand new `Datadog::Tracer` instance from scratch.

I've seen clients create custom tracers for testing, and they normally overwrite one method or two.

File: lib/ddtrace/configuration/components.rb (hunk 1 above)
Comment (ivoanjo) [reply lib/ddtrace/configuration/components.rb:93]: > Because they can be completely overwritten by the user with custom instances, it makes sense for them to be easy to track, and not overly attached to our tracing internals.

Ah. This is a pretty good reason. I somewhat question _that_ decision, but I feel that ship has sailed, so given that constraint, I see the point of giving them a more important standing.

File: lib/ddtrace/configuration/components.rb (hunk 5 above)
Comment (ivoanjo) [reply lib/ddtrace/configuration/components.rb]: > The issue is that we told plenty of clients to use (...)

Ahh, yes, reality sets in. Agree with you here, probably not worth the user pain just so that we can "make it prettier" on our side.

File: lib/ddtrace/context_flush.rb (hunk 6 above)
Comment (ivoanjo) [reply lib/ddtrace/context_flush.rb:30]: I see. I think having it private is a bit better -- it's sometimes "dirtier" and a bit of a code smell to be testing private methods directly, but between that and clearly signaling what customers should and shouldn't use, I think the latter ends up being more important.

> I don't think we have the right tools in Ruby sometimes for testing.

Agreed.

File: lib/ddtrace/configuration/components.rb (hunk 3 above)
Comment (ivoanjo) [reply lib/ddtrace/configuration/components.rb:68]: > Break off the initialization changes of the Writer and Sampler from removing Tracer#configure?
> 
> I could do the initialization changes of the Writer and Sampler before removing Tracer#configure, but not in parallel. They would also overlap a bit: I'd have to write a stop-gap solution for the Writer#update_priority_sampler change, reconciling the old way of providing the priority_sampler to the Writer, while maintaining it at Tracer#configure and write tests to account for both cases and their conflicts, and then remove that stop-gap solution in the Tracer#configure PR.

Hmmm this may bit more complex than what I was thinking, but that wasn't quite what I was suggesting.

The thing I'm not quite convinced is the right direction and would like to look at separately is the introduction of events into `Writer` and `SyncWriter` as an indirect way of connecting them to the `Sampler`.

That's because I don't see it as a very generic abstraction that we'll end up using for lots of things, but as a generic abstraction which we'll use one for this one thing -- connecting the writers to the sampler -- and thus I fear it'll be one more layer of indirection added to something that's already very complex and hard to follow.

**Comment by ivoanjo** [comment]
Overall this seems reasonable, but I think given some of our discussions/questions, it probably makes sense to have @delner take a look and give the final approval.

File: lib/ddtrace/opentracer/tracer.rb [hunk 9]
```diff
@@ -1,4 +1,4 @@
-# typed: true
+# typed: false
```
Comment (marcotc) [inline lib/ddtrace/opentracer/tracer.rb:1]:
```
lib/ddtrace/opentracer/tracer.rb:19: Passing a hash where the specific keys are unknown to a method taking keyword arguments https://srb.help/7019
    19 |        @datadog_tracer = Datadog::Tracer.new(**options)
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It's asking me to add all `Datadog::Tracer#initialize` arguments to this class, which I don't think is a good idea.
I couldn't find a way to disable Sorbet only for the `Datadog::OpenTracer::Tracer#initialize` method, so I had to disable it for the whole file 😢

Comment (ivoanjo) [reply lib/ddtrace/opentracer/tracer.rb:1]: Yup. To be honest, I'm not sure Sorbet has been adding a lot of value since its initial addition (where it caught a few existing issues), so I guess once we add support for steep, it may make sense to touch base and decide what to do with Sorbet.

Comment (marcotc) [reply lib/ddtrace/opentracer/tracer.rb:1]: It did find at least a couple issues for me during development since the introduction.

Comment (delner) [reply lib/ddtrace/opentracer/tracer.rb:1]: Sorbet has cost me more time than it's saved, and it doesn't have a good means of ignoring things I know are okay, forcing me to ignore the whole file.

A discussion for another thread, but this is yet another exhibit for the case against it later.

**Review by delner — APPROVED** [review]
We reviewed this together, looks good! Thank you!
